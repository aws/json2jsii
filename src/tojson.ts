import { Code } from './code';

export class toJsonFunction {
  public static nameOf(typeName: string) {
    return `${typeName}$toJson`;
  }

  private readonly fields: Record<string, string> = {};

  public readonly functionName: string;

  constructor(private readonly baseType: string) {
    this.functionName = toJsonFunction.nameOf(this.baseType);
  }

  public addField(schemaName: string, propertyName: string, toJson: toJson) {
    this.fields[schemaName] = toJson(`obj.${propertyName}`);
  }

  public emit(code: Code) {
    code.line();
    code.line('/**');
    code.line(` * Converts an object of type '${this.baseType}' to JSON representation.`);
    code.line(' */');
    code.openBlock(`export function ${this.functionName}(obj: ${this.baseType} | undefined): Record<string, any> | undefined`);
    code.line('if (obj === undefined) { return undefined; }');
    code.open('const result = {');
    for (const [k, v] of Object.entries(this.fields)) {
      code.line(`'${k}': ${v},`);
    }
    code.close('};');

    code.line('// filter undefined values');
    code.line('return Object.entries(result).reduce((r, i) => (i[1] === undefined) ? r : ({ ...r, [i[0]]: i[1] }), {});');

    code.closeBlock();
  }
}

export type toJson = (x: string) => string;