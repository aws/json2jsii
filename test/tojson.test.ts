
import { spawnSync } from 'child_process';
import { mkdtempSync, writeFileSync } from 'fs';
import { tmpdir } from 'os';
import { join } from 'path';
import { JSONSchema4 } from 'json-schema';
import { TypeGenerator } from '../src';
import { toJsonFunction } from '../src/jsonify';

test('primitives', () => {
  const f = compile({
    properties: {
      StringProperty: { type: 'string' },
      BooleanProperty: { type: 'boolean' },
      NumberProperty: { type: 'number' },
      With_UnderScore: { type: 'integer' },
    },
    required: [
      'StringProperty',
    ],
  });

  expect(f({
    stringProperty: 'hello', booleanProperty: false, numberProperty: 1234, withUnderScore: 8989,
  })).toStrictEqual({
    StringProperty: 'hello', BooleanProperty: false, NumberProperty: 1234, With_UnderScore: 8989,
  });

  expect(f({
    stringProperty: 'hello, world', numberProperty: undefined,
  })).toStrictEqual({
    StringProperty: 'hello, world',
  });
});

test('names can get crazy', () => {
  const f = compile({
    properties: {
      'hyphen-name': { type: 'string' },
      '$nameWithDolar': { type: 'string' },
      'name with spaces': { type: 'string' },
    },
  });

  expect(f({
    hyphenName: 'boomboom',
    nameWithDolar: 'hey',
    nameWithSpaces: 'glow',
  })).toStrictEqual({
    'hyphen-name': 'boomboom',
    '$nameWithDolar': 'hey',
    'name with spaces': 'glow',
  });
});

test('complex types', () => {
  const f = compile({
    definitions: {
      YourType: {
        properties: {
          John: { type: 'array', items: { type: 'string' } },
        },
      },
      MyType: {
        properties: {
          Foo: { type: 'number' },
          Bar: { type: 'string' },
          Nested: { $ref: '#/definitions/YourType' },
        },
        required: ['Nested'],
      },
    },
    properties: {
      ComplexType: { $ref: '#/definitions/MyType' },
    },
  });

  expect(f({
    complexType: { foo: 1234, bar: 'hey there', nested: { john: ['foo'] } },
  })).toStrictEqual({
    ComplexType: { Foo: 1234, Bar: 'hey there', Nested: { John: ['foo'] } },
  });
});

describe('arrays', () => {

  test('of primitives', () => {
    const f = compile({
      properties: {
        ArrayProp: { type: 'array', items: { type: 'string' } },
      },
    });

    expect(f({
      arrayProp: ['bello', 'aoll'],
    })).toStrictEqual({
      ArrayProp: ['bello', 'aoll'],
    });
  });

  test('of complex types', () => {
    const f = compile({
      definitions: {
        MyType: {
          properties: {
            Foo: { type: 'number' },
            Bar: { type: 'string' },
          },
        },
      },
      properties: {
        Array_Of_complex: { type: 'array', items: { $ref: '#/definitions/MyType' } },
      },
    });

    expect(f({
      arrayOfComplex: [{ foo: 122 }, { bar: 'hello' }, { foo: 88, bar: 'world' }],
    })).toStrictEqual({
      Array_Of_complex: [{ Foo: 122 }, { Bar: 'hello' }, { Foo: 88, Bar: 'world' }],
    });
  });

});

test('any', () => {
  const f = compile({
    properties: {
      MyAny: { type: 'object' },
    },
  });

  expect(f({ myAny: 177 })).toStrictEqual({ MyAny: 177 });
  expect(f({ myAny: 'hello' })).toStrictEqual({ MyAny: 'hello' });
  expect(f({ myAny: ['hello', 889] })).toStrictEqual({ MyAny: ['hello', 889] });
});


describe('maps', () => {
  test('of primitives', () => {
    const f = compile({
      properties: {
        StringMap: { additionalProperties: { type: 'string' } },
        NumberMap: { additionalProperties: { type: 'number' } },
      },
    });

    expect(f({
      stringMap: {
        hello: 'string',
        world: 'strong',
      },
      numberMap: {
        hello: 10,
        world: 20,
      },
    })).toStrictEqual({
      StringMap: {
        hello: 'string',
        world: 'strong',
      },
      NumberMap: {
        hello: 10,
        world: 20,
      },
    });

    expect(f({
      stringMap: undefined,
      numberMap: {
        hello: undefined,
        world: 11,
      },
    })).toStrictEqual({
      NumberMap: {
        world: 11,
      },
    });
  });

  test('of complex', () => {
    const f = compile({
      definitions: {
        MyType: {
          properties: {
            Foo: { type: 'number' },
            Bar: { type: 'string' },
          },
        },
      },
      properties: {
        ComplexMap: { additionalProperties: { $ref: '#/definitions/MyType' } },
      },
    });

    expect(f({
      complexMap: {
        foya: { foo: 123, bar: 'barbar' },
        hello_world: { foo: 3333 },
      },
    })).toStrictEqual({
      ComplexMap: {
        foya: { Foo: 123, Bar: 'barbar' },
        hello_world: { Foo: 3333 },
      },
    });
  });
});

test('enums', () => {
  const f = compile({
    properties: {
      MyEnum: { enum: ['one', 'two', 'three'] },
      YourEnum: { enum: ['jo', 'shmo'] },
    },
  });

  expect(f({
    myEnum: 'two',
    yourEnum: undefined,
  })).toStrictEqual({
    MyEnum: 'two',
  });
});

test('date', () => {
  const f = compile({
    properties: {
      DateTime: { type: 'string', format: 'date-time' },
    },
  });

  expect(f({
    dateTime: new Date('2021-07-19T12:08:52.210Z'),
  })).toStrictEqual({
    DateTime: '2021-07-19T12:08:52.210Z',
  });

  expect(f({ dateTime: undefined })).toStrictEqual({});
});

/**
 * Generates structs from this JSON schema and returns a compiled versio of the
 * toJson method for the top-level struct.
 */
function compile(schema: JSONSchema4): (x: any) => any {
  // generate code into workdir
  const workdir = mkdtempSync(join(tmpdir(), 'tojson.'));
  const gen = TypeGenerator.forStruct('MyStruct', schema);
  const code = gen.render();
  expect(code).toMatchSnapshot();

  writeFileSync(join(workdir, 'index.ts'), code);

  try {
    // compile using the typescript compiler
    const tsc = require.resolve('typescript/bin/tsc');
    spawnSync(tsc, ['index.ts'], { cwd: workdir });

    // import the compiled javascript code into this process (wow)
    const source = join(workdir, 'index.js');

    const fnName = toJsonFunction.nameOf('MyStruct');

    // eslint-disable-next-line @typescript-eslint/no-require-imports
    return require(source)[fnName];
  } catch (e) {
    throw new Error(`Compilation error: ${e}. Workdir: ${workdir}`);
  }
}