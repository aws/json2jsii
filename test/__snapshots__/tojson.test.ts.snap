// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`any 1`] = `
"/**
 * @schema MyStruct
 */
export interface MyStruct {
  /**
   * @schema MyStruct#MyAny
   */
  readonly myAny?: any;

}

/**
 * Converts an object of type 'MyStruct' to JSON representation.
 */
export function MyStruct$toJson(obj: MyStruct | undefined): Record<string, any> | undefined {
  if (obj === undefined) { return undefined; }
  const result = {
    'MyAny': obj.myAny,
  };
  // filter undefined values
  return Object.entries(result).reduce((r, i) => (i[1] === undefined) ? r : ({ ...r, [i[0]]: i[1] }), {});
}
"
`;

exports[`arrays of complex types 1`] = `
"/**
 * @schema MyStruct
 */
export interface MyStruct {
  /**
   * @schema MyStruct#Array_Of_complex
   */
  readonly arrayOfComplex?: MyType[];

}

/**
 * Converts an object of type 'MyStruct' to JSON representation.
 */
export function MyStruct$toJson(obj: MyStruct | undefined): Record<string, any> | undefined {
  if (obj === undefined) { return undefined; }
  const result = {
    'Array_Of_complex': obj.arrayOfComplex?.map(y => MyType$toJson(y)),
  };
  // filter undefined values
  return Object.entries(result).reduce((r, i) => (i[1] === undefined) ? r : ({ ...r, [i[0]]: i[1] }), {});
}

/**
 * @schema MyType
 */
export interface MyType {
  /**
   * @schema MyType#Foo
   */
  readonly foo?: number;

  /**
   * @schema MyType#Bar
   */
  readonly bar?: string;

}

/**
 * Converts an object of type 'MyType' to JSON representation.
 */
export function MyType$toJson(obj: MyType | undefined): Record<string, any> | undefined {
  if (obj === undefined) { return undefined; }
  const result = {
    'Foo': obj.foo,
    'Bar': obj.bar,
  };
  // filter undefined values
  return Object.entries(result).reduce((r, i) => (i[1] === undefined) ? r : ({ ...r, [i[0]]: i[1] }), {});
}
"
`;

exports[`arrays of primitives 1`] = `
"/**
 * @schema MyStruct
 */
export interface MyStruct {
  /**
   * @schema MyStruct#ArrayProp
   */
  readonly arrayProp?: string[];

}

/**
 * Converts an object of type 'MyStruct' to JSON representation.
 */
export function MyStruct$toJson(obj: MyStruct | undefined): Record<string, any> | undefined {
  if (obj === undefined) { return undefined; }
  const result = {
    'ArrayProp': obj.arrayProp?.map(y => y),
  };
  // filter undefined values
  return Object.entries(result).reduce((r, i) => (i[1] === undefined) ? r : ({ ...r, [i[0]]: i[1] }), {});
}
"
`;

exports[`complex types 1`] = `
"/**
 * @schema MyStruct
 */
export interface MyStruct {
  /**
   * @schema MyStruct#ComplexType
   */
  readonly complexType?: MyType;

}

/**
 * Converts an object of type 'MyStruct' to JSON representation.
 */
export function MyStruct$toJson(obj: MyStruct | undefined): Record<string, any> | undefined {
  if (obj === undefined) { return undefined; }
  const result = {
    'ComplexType': MyType$toJson(obj.complexType),
  };
  // filter undefined values
  return Object.entries(result).reduce((r, i) => (i[1] === undefined) ? r : ({ ...r, [i[0]]: i[1] }), {});
}

/**
 * @schema MyType
 */
export interface MyType {
  /**
   * @schema MyType#Foo
   */
  readonly foo?: number;

  /**
   * @schema MyType#Bar
   */
  readonly bar?: string;

  /**
   * @schema MyType#Nested
   */
  readonly nested: YourType;

}

/**
 * Converts an object of type 'MyType' to JSON representation.
 */
export function MyType$toJson(obj: MyType | undefined): Record<string, any> | undefined {
  if (obj === undefined) { return undefined; }
  const result = {
    'Foo': obj.foo,
    'Bar': obj.bar,
    'Nested': YourType$toJson(obj.nested),
  };
  // filter undefined values
  return Object.entries(result).reduce((r, i) => (i[1] === undefined) ? r : ({ ...r, [i[0]]: i[1] }), {});
}

/**
 * @schema YourType
 */
export interface YourType {
  /**
   * @schema YourType#John
   */
  readonly john?: string[];

}

/**
 * Converts an object of type 'YourType' to JSON representation.
 */
export function YourType$toJson(obj: YourType | undefined): Record<string, any> | undefined {
  if (obj === undefined) { return undefined; }
  const result = {
    'John': obj.john?.map(y => y),
  };
  // filter undefined values
  return Object.entries(result).reduce((r, i) => (i[1] === undefined) ? r : ({ ...r, [i[0]]: i[1] }), {});
}
"
`;

exports[`date 1`] = `
"/**
 * @schema MyStruct
 */
export interface MyStruct {
  /**
   * @schema MyStruct#DateTime
   */
  readonly dateTime?: Date;

}

/**
 * Converts an object of type 'MyStruct' to JSON representation.
 */
export function MyStruct$toJson(obj: MyStruct | undefined): Record<string, any> | undefined {
  if (obj === undefined) { return undefined; }
  const result = {
    'DateTime': obj.dateTime?.toISOString(),
  };
  // filter undefined values
  return Object.entries(result).reduce((r, i) => (i[1] === undefined) ? r : ({ ...r, [i[0]]: i[1] }), {});
}
"
`;

exports[`enums 1`] = `
"/**
 * @schema MyStruct
 */
export interface MyStruct {
  /**
   * @schema MyStruct#MyEnum
   */
  readonly myEnum?: MyStructMyEnum;

  /**
   * @schema MyStruct#YourEnum
   */
  readonly yourEnum?: MyStructYourEnum;

}

/**
 * Converts an object of type 'MyStruct' to JSON representation.
 */
export function MyStruct$toJson(obj: MyStruct | undefined): Record<string, any> | undefined {
  if (obj === undefined) { return undefined; }
  const result = {
    'MyEnum': obj.myEnum,
    'YourEnum': obj.yourEnum,
  };
  // filter undefined values
  return Object.entries(result).reduce((r, i) => (i[1] === undefined) ? r : ({ ...r, [i[0]]: i[1] }), {});
}

/**
 * @schema MyStructMyEnum
 */
export enum MyStructMyEnum {
  /** one */
  ONE = \\"one\\",
  /** two */
  TWO = \\"two\\",
  /** three */
  THREE = \\"three\\",
}

/**
 * @schema MyStructYourEnum
 */
export enum MyStructYourEnum {
  /** jo */
  JO = \\"jo\\",
  /** shmo */
  SHMO = \\"shmo\\",
}
"
`;

exports[`maps of complex 1`] = `
"/**
 * @schema MyStruct
 */
export interface MyStruct {
  /**
   * @schema MyStruct#ComplexMap
   */
  readonly complexMap?: { [key: string]: MyType };

}

/**
 * Converts an object of type 'MyStruct' to JSON representation.
 */
export function MyStruct$toJson(obj: MyStruct | undefined): Record<string, any> | undefined {
  if (obj === undefined) { return undefined; }
  const result = {
    'ComplexMap': ((obj.complexMap) === undefined) ? undefined : (Object.entries(obj.complexMap).reduce((r, i) => (i[1] === undefined) ? r : ({ ...r, [i[0]]: MyType$toJson(i[1]) }), {})),
  };
  // filter undefined values
  return Object.entries(result).reduce((r, i) => (i[1] === undefined) ? r : ({ ...r, [i[0]]: i[1] }), {});
}

/**
 * @schema MyType
 */
export interface MyType {
  /**
   * @schema MyType#Foo
   */
  readonly foo?: number;

  /**
   * @schema MyType#Bar
   */
  readonly bar?: string;

}

/**
 * Converts an object of type 'MyType' to JSON representation.
 */
export function MyType$toJson(obj: MyType | undefined): Record<string, any> | undefined {
  if (obj === undefined) { return undefined; }
  const result = {
    'Foo': obj.foo,
    'Bar': obj.bar,
  };
  // filter undefined values
  return Object.entries(result).reduce((r, i) => (i[1] === undefined) ? r : ({ ...r, [i[0]]: i[1] }), {});
}
"
`;

exports[`maps of primitives 1`] = `
"/**
 * @schema MyStruct
 */
export interface MyStruct {
  /**
   * @schema MyStruct#StringMap
   */
  readonly stringMap?: { [key: string]: string };

  /**
   * @schema MyStruct#NumberMap
   */
  readonly numberMap?: { [key: string]: number };

}

/**
 * Converts an object of type 'MyStruct' to JSON representation.
 */
export function MyStruct$toJson(obj: MyStruct | undefined): Record<string, any> | undefined {
  if (obj === undefined) { return undefined; }
  const result = {
    'StringMap': ((obj.stringMap) === undefined) ? undefined : (Object.entries(obj.stringMap).reduce((r, i) => (i[1] === undefined) ? r : ({ ...r, [i[0]]: i[1] }), {})),
    'NumberMap': ((obj.numberMap) === undefined) ? undefined : (Object.entries(obj.numberMap).reduce((r, i) => (i[1] === undefined) ? r : ({ ...r, [i[0]]: i[1] }), {})),
  };
  // filter undefined values
  return Object.entries(result).reduce((r, i) => (i[1] === undefined) ? r : ({ ...r, [i[0]]: i[1] }), {});
}
"
`;

exports[`names can get crazy 1`] = `
"/**
 * @schema MyStruct
 */
export interface MyStruct {
  /**
   * @schema MyStruct#hyphen-name
   */
  readonly hyphenName?: string;

  /**
   * @schema MyStruct#$nameWithDolar
   */
  readonly nameWithDolar?: string;

  /**
   * @schema MyStruct#name with spaces
   */
  readonly nameWithSpaces?: string;

}

/**
 * Converts an object of type 'MyStruct' to JSON representation.
 */
export function MyStruct$toJson(obj: MyStruct | undefined): Record<string, any> | undefined {
  if (obj === undefined) { return undefined; }
  const result = {
    'hyphen-name': obj.hyphenName,
    '$nameWithDolar': obj.nameWithDolar,
    'name with spaces': obj.nameWithSpaces,
  };
  // filter undefined values
  return Object.entries(result).reduce((r, i) => (i[1] === undefined) ? r : ({ ...r, [i[0]]: i[1] }), {});
}
"
`;

exports[`primitives 1`] = `
"/**
 * @schema MyStruct
 */
export interface MyStruct {
  /**
   * @schema MyStruct#StringProperty
   */
  readonly stringProperty: string;

  /**
   * @schema MyStruct#BooleanProperty
   */
  readonly booleanProperty?: boolean;

  /**
   * @schema MyStruct#NumberProperty
   */
  readonly numberProperty?: number;

  /**
   * @schema MyStruct#With_UnderScore
   */
  readonly withUnderScore?: number;

}

/**
 * Converts an object of type 'MyStruct' to JSON representation.
 */
export function MyStruct$toJson(obj: MyStruct | undefined): Record<string, any> | undefined {
  if (obj === undefined) { return undefined; }
  const result = {
    'StringProperty': obj.stringProperty,
    'BooleanProperty': obj.booleanProperty,
    'NumberProperty': obj.numberProperty,
    'With_UnderScore': obj.withUnderScore,
  };
  // filter undefined values
  return Object.entries(result).reduce((r, i) => (i[1] === undefined) ? r : ({ ...r, [i[0]]: i[1] }), {});
}
"
`;
